mava/systems/ff_ippo.py

```diff
@@ -399,71 +401,74 @@ def learner_setup(
-
-    # Add dimension to pmap over.
-    reshape_step_rngs = lambda x: x.reshape(
-        (n_devices, config.system.update_batch_size) + x.shape[1:]
-    )
-    step_rngs = reshape_step_rngs(jnp.stack(step_rngs))
     reshape_states = lambda x: x.reshape(
         (n_devices, config.system.update_batch_size, config.arch.num_envs) + x.shape[1:]
     )
-    env_states = jax.tree_util.tree_map(reshape_states, env_states)
-    timesteps = jax.tree_util.tree_map(reshape_states, timesteps)
+    # (devices, update batch size, num_envs, ...)
+    env_states = jax.tree_map(reshape_states, env_states)
+    timesteps = jax.tree_map(reshape_states, timesteps)
 
+    # Define params to be replicated across devices and batches.
+    key, step_keys = jax.random.split(key)
     params = Params(actor_params, critic_params)
     opt_states = OptStates(actor_opt_state, critic_opt_state)
+    replicate_learner = (params, opt_states, step_keys)
+
+    # Duplicate learner for update_batch_size.
+    broadcast = lambda x: jnp.broadcast_to(x, (config.system.update_batch_size,) + x.shape)
+    replicate_learner = jax.tree_map(broadcast, replicate_learner)
+
+    # Duplicate learner across devices.
+    replicate_learner = flax.jax_utils.replicate(replicate_learner, devices=jax.devices())
+
+    # Initialise learner state.
+    params, opt_states, step_keys = replicate_learner
+    init_learner_state = LearnerState(params, opt_states, step_keys, env_states, timesteps)
 
-    init_learner_state = LearnerState(params, opt_states, step_rngs, env_states, timesteps)
     return learn, actor_network, init_learner_state
 
 
 def run_experiment(_config: DictConfig) -> None:
     """Runs experiment."""
-    # Logger setup
     config = copy.deepcopy(_config)
-    log = logger_setup(config)
 
     # Create the enviroments for train and eval.
-    env, eval_env = make(config=config)
+    env, eval_env = environments.make(config)
 
     # PRNG keys.
-    rng, rng_e, rng_p = jax.random.split(jax.random.PRNGKey(config.system.seed), num=3)
+    key, key_e, actor_net_key, critic_net_key = jax.random.split(
+        jax.random.PRNGKey(config["system"]["seed"]), num=4
+    )
 
     # Setup learner.
-    learn, actor_network, learner_state = learner_setup(env, (rng, rng_p), config)
+    learn, actor_network, learner_state = learner_setup(
+        env, (key, actor_net_key, critic_net_key), config
+    )
 
     # Setup evaluator.
-    evaluator, absolute_metric_evaluator, (trained_params, eval_rngs) = evaluator_setup(
+    evaluator, absolute_metric_evaluator, (trained_params, eval_keys) = evaluator_setup(
         eval_env=eval_env,
-        rng_e=rng_e,
+        key_e=key_e,
         network=actor_network,
         params=learner_state.params.actor_params,
         config=config,
     )
 
     # Calculate total timesteps.
     n_devices = len(jax.devices())
+    config = check_total_timesteps(config)
+    assert (
```

Comment:
Can we use raise errors instead? Since it can be ignored with `python -O file.py` command and someone who's not familiar with **Mava**  he'll receive an error at the start of the evaluation phase without knowing the root cause 🤔

Reply:
`assert` is more appropriate for this to be honest, happy to have a private chat with you about it